# Capstone Project - Toronto Neighborhood

First we import all the required libraries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
from IPython.display import display_html

    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

# plotting library
import folium 
from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

Next, we download the neighborhood information available on wikipedia

In [2]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup=BeautifulSoup(source,'lxml')
print(soup.title)
from IPython.display import display_html
tab = str(soup.table)
display_html(tab,raw=True)

<title>List of postal codes of Canada: M - Wikipedia</title>


M1A Not assigned 
 
 
 
 M2A Not assigned 
 
 
 
 M3A North York ( Parkwoods ) 
 
 
 
 M4A North York ( Victoria Village ) 
 
 
 
 M5A Downtown Toronto ( Regent Park / Harbourfront ) 
 
 
 
 M6A North York ( Lawrence Manor / Lawrence Heights ) 
 
 
 
 M7A Queen's Park (Ontario Provincial Government) 
 
 
 
 M8A Not assigned 
 
 
 
 M9A Etobicoke ( Islington Avenue ) 
 
 
 
 
 M1B Scarborough ( Malvern / Rouge ) 
 
 
 
 M2B Not assigned 
 
 
 
 M3B North York ( Don Mills ) North 
 
 
 
 M4B East York ( Parkview Hill / Woodbine Gardens ) 
 
 
 
 M5B Downtown Toronto ( Garden District , Ryerson ) 
 
 
 
 M6B North York (Glencairn) 
 
 
 
 M7B Not assigned 
 
 
 
 M8B Not assigned 
 
 
 
 M9B Etobicoke ( West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale) 
 
 
 
 
 M1C Scarborough ( Rouge Hill / Port Union / Highland Creek ) 
 
 
 
 M2C Not assigned 
 
 
 
 M3C North York ( Don Mills ) South ( Flemingdon Park ) 
 
 
 
 M4C East York ( Woodbine Heights ) 
 
 
 
 M5C Downtown Toronto ( St. James Town ) 
 
 
 
 M6C York ( Humewood-Cedarvale ) 
 
 
 
 M7C Not assigned 
 
 
 
 M8C Not assigned 
 
 
 
 M9C Etobicoke (Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood ) 
 
 
 
 
 M1E Scarborough ( Guildwood / Morningside / West Hill ) 
 
 
 
 M2E Not assigned 
 
 
 
 M3E Not assigned 
 
 
 
 M4E East Toronto ( The Beaches ) 
 
 
 
 M5E Downtown Toronto (Berczy Park) 
 
 
 
 M6E York ( Caledonia-Fairbanks ) 
 
 
 
 M7E Not assigned 
 
 
 
 M8E Not assigned 
 
 
 
 M9E Not assigned 
 
 
 
 
 M1G Scarborough ( Woburn ) 
 
 
 
 M2G Not assigned 
 
 
 
 M3G Not assigned 
 
 
 
 M4G East York ( Leaside ) 
 
 
 
 M5G Downtown Toronto (Central Bay Street ) 
 
 
 
 M6G Downtown Toronto (Christie) 
 
 
 
 M7G Not assigned 
 
 
 
 M8G Not assigned 
 
 
 
 M9G Not assigned 
 
 
 
 
 M1H Scarborough ( Cedarbrae ) 
 
 
 
 M2H North York ( Hillcrest Village ) 
 
 
 
 M3H North York ( Bathurst Manor / Wilson Heights / Downsview North) 
 
 
 
 M4H East York ( Thorncliffe Park ) 
 
 
 
 M5H Downtown Toronto ( Richmond / Adelaide / King ) 
 
 
 
 M6H West Toronto ( Dufferin / Dovercourt Village ) 
 
 
 
 M7H Not assigned 
 
 
 
 M8H Not assigned 
 
 
 
 M9H Not assigned 
 
 
 
 
 M1J Scarborough ( Scarborough Village ) 
 
 
 
 M2J North York (Fairview / Henry Farm / Oriole) 
 
 
 
 M3J North York ( Northwood Park / York University ) 
 
 
 
 M4J East York East Toronto ( The Danforth East) 
 
 
 
 M5J Downtown Toronto ( Harbourfront East / Union Station / Toronto Islands ) 
 
 
 
 M6J West Toronto ( Little Portugal / Trinity ) 
 
 
 
 M7J Not assigned 
 
 
 
 M8J Not assigned 
 
 
 
 M9J Not assigned 
 
 
 
 
 M1K Scarborough ( Kennedy Park / Ionview / East Birchmount Park ) 
 
 
 
 M2K North York ( Bayview Village ) 
 
 
 
 M3K North York ( Downsview ) East ( CFB Toronto ) 
 
 
 
 M4K East Toronto ( The Danforth West / Riverdale ) 
 
 
 
 M5K Downtown Toronto ( Toronto Dominion Centre / Design Exchange ) 
 
 
 
 M6K West Toronto ( Brockton / Parkdale Village / Exhibition Place ) 
 
 
 
 M7K Not assigned 
 
 
 
 M8K Not assigned 
 
 
 
 M9K Not assigned 
 
 
 
 
 M1L Scarborough ( Golden Mile / Clairlea / Oakridge ) 
 
 
 
 M2L North York ( York Mills / Silver Hills) 
 
 
 
 M3L North York ( Downsview ) West 
 
 
 
 M4L East Toronto ( India Bazaar / The Beaches West) 
 
 
 
 M5L Downtown Toronto ( Commerce Court / Victoria Hotel ) 
 
 
 
 M6L North York (North Park / Maple Leaf Park / Upwood Park) 
 
 
 
 M7L Not assigned 
 
 
 
 M8L Not assigned 
 
 
 
 M9L North York ( Humber Summit ) 
 
 
 
 
 M1M Scarborough ( Cliffside / Cliffcrest / Scarborough Village West) 
 
 
 
 M2M North York (Willowdale / Newtonbrook ) 
 
 
 
 M3M North York ( Downsview ) Central 
 
 
 
 M4M East Toronto (Studio District) 
 
 
 
 M5M North York ( Bedford Park / Lawrence Manor East) 
 
 
 
 M6M York (Del Ray / Mount Dennis / Keelsdale and Silverthorn ) 
 
 
 
 M7M Not assigned 
 
 
 
 M8M Not assigned 
 
 
 
 M9M North York ( Humberlea / Emery

As can be seen above, the neighborhood information is currently grouped as Postal Code, Borough and Neighborhood. In the next cell, we split this information and assigned into proper columns of a database.

In [3]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)



We next print the database after rewriting some information in Borough column for better readibility.

In [4]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [5]:
df.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                    Don Mills South  
14                                   Woodbine Heights

In [6]:
df.shape

(103, 3)

In [7]:
lat_lon = pd.read_csv('https://cocl.us/Geospatial_data')
lat_lon.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Merging the two tables for getting the Latitudes and Longitudes for various neighbourhoods in Canada

In [8]:
lat_lon.rename(columns={'Postal Code':'PostalCode'},inplace=True)
df2 = pd.merge(df,lat_lon,on='PostalCode')
df2.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [9]:
df2.shape

(103, 5)

## 3. Exploring and clustering the neighborhoods in Toronto

Isolating boroughs containing 'Toronto' in the name

In [10]:
df3 = df2[df2['Borough'].str.contains('Toronto',regex=False)]
df3

PostalCode                 Borough  \
2          M5A        Downtown Toronto   
9          M5B        Downtown Toronto   
15         M5C        Downtown Toronto   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
54         M4M            East Toronto   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
96         M4X        Downtown Toronto   
97         M5X        Downtown Toronto   
99         M4Y        Downtown Toronto   
100        M7Y   East Toronto Business   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
35                                  The Danforth  East  43.685347 -79.338106  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68                            Forest Hill North & West  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                                  North Toronto West  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                              

In [11]:
df3.shape

(39, 5)

The number of rows have reduced from the previous 103 to 39 when boroughs are filtered for names having Toronto.

### Visualizing all Toronto Neighbourhoods using Folium

Next we use Folium to generate maps and visualize Toronto neighborhood.

In [12]:
map_toronto = folium.Map(location=[43.651070,-79.347015],zoom_start=10)

for lat,lng,borough,neighbourhood in zip(df3['Latitude'],df3['Longitude'],df3['Borough'],df3['Neighborhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
map_toronto